pkg/planner/core/logical_plan_builder.go

```diff
@@ -5718,13 +5721,109 @@ func buildColumns2Handle(
 				return nil, err
 			}
 			end := offset + tblLen
-			cols2Handles = append(cols2Handles, TblColPosInfo{tblID, offset, end, handleCol})
+			cols2Handles = append(cols2Handles, TblColPosInfo{TblID: tblID, Start: offset, End: end, HandleCols: handleCol})
 		}
 	}
 	sort.Sort(cols2Handles)
 	return cols2Handles, nil
 }
 
+// buildColPositionInfoForDelete builds columns to handle mapping for delete.
+// We'll have two kinds of columns seen by DELETE:
+//  1. The columns that are public. They are the columns that not affected by any DDL.
+//  2. The columns that are not public. They are the columns that are affected by DDL.
+//     But we need them because the non-public indexes may rely on them.
+//
+// The two kind of columns forms the whole columns of the table. Public part first.
+// This function records the following things:
+//  1. The position of the columns used by indexes in the DELETE's select's output.
+//  2. The row id's position in the output.
+//  3. The possible reference column of the column under MODIFY/CHANGE ddl job.
+func buildColPositionInfoForDelete(
+	names []*types.FieldName,
+	tblID2Handle map[int64][]util.HandleCols,
+	tblID2Table map[int64]table.Table,
+) (TblColPosInfoSlice, error) {
+	var cols2PosInfos TblColPosInfoSlice
+	for tblID, handleCols := range tblID2Handle {
+		tbl := tblID2Table[tblID]
+		pubCols := tbl.Cols()
+		deletableIdxs := tbl.DeletableIndices()
+		deletableCols := tbl.DeletableCols()
+		tblInfo := tbl.Meta()
+
+		for _, handleCol := range handleCols {
+			curColPosInfo, err := buildSingleTableColPosInfoForDelete(names, handleCol, pubCols, deletableIdxs, deletableCols, tblInfo)
+			if err != nil {
+				return nil, err
+			}
+			cols2PosInfos = append(cols2PosInfos, curColPosInfo)
+		}
+	}
+	sort.Sort(cols2PosInfos)
+	return cols2PosInfos, nil
+}
+
+// buildSingleTableColPosInfoForDelete builds columns mapping for delete.
+func buildSingleTableColPosInfoForDelete(
+	names []*types.FieldName,
+	handleCol util.HandleCols,
+	pubCols []*table.Column,
+	deletableIdxs []table.Index,
+	deletableCols []*table.Column,
+	tblInfo *model.TableInfo,
+) (TblColPosInfo, error) {
+	// Columns can be seen by DELETE are the deletable columns.
+	tblLen := len(deletableCols)
+	offset, err := getTableOffset(names, names[handleCol.GetCol(0).Index])
+	if err != nil {
+		return TblColPosInfo{}, err
+	}
+	end := offset + tblLen
+
+	// Index only records its columns' offsets in the deletableCols(the whole columns).
+	// So we need to first change the offsets to the real position of SELECT's output.
+	offsetMap := make(map[int]int, len(deletableCols))
+	for i := offset; i < end; i++ {
+		name := names[i]
+		found := -1
+		for j, col := range deletableCols {
+			if col.Name.L == name.ColName.L {
+				offsetMap[j] = i
+				found = j
+				break
+			}
+		}
+		if found == -1 {
+			return TblColPosInfo{}, plannererrors.ErrDeleteNotFoundColumn.GenWithStackByArgs(name.ColName.O, tblInfo.Name.O)
```

Comment:
Could the util method like `FindTableByName` be used here?

Reply:
No, most common cases are focused on `*model.ColumnInfo`, not `*table.Column`